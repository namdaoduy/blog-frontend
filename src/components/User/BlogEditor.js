import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import 'draft-js/dist/Draft.css';
import React, { Component } from 'react';
import ReactMde from 'react-mde';
import 'react-mde/lib/styles/css/react-mde-all.css';
import { connect } from 'react-redux';
import * as Showdown from 'showdown';
import { createBlog, getEdittingBlog, updateBlog } from '../../actions/user';
import '../../assets/styles/user.css';
import * as constants from '../../constants/common';
import history from '../../utils/history';
import Header from '../Common/Header';

class BlogEditor extends Component {
  constructor(props) {
    super(props);
    const blog_id = props.match.params.blog_id;
    this.state = {
      id: blog_id || null,
      type: blog_id ? 'edit' : 'new',
      title: '',
      body: 'Your body here',
      isLoading: true,
      error: false,
    };

    this.converter = new Showdown.Converter({
      tables: true,
      simplifiedAutoLink: true,
    });
  }

  componentDidMount() {
    const { id, type } = this.state;
    if (type === 'new') {
      this.setState({
        isLoading: false,
        title: '',
        body: 'Your body here',
      });
    } else {
      this.fetchBlog(id);
    }
  }

  handleBodyChange = (body) => {
    this.setState({ body });
  }

  handleTitleChange = (e) => {
    this.setState({ title: e.target.value });
  }

  // If create new blog
  handleNew = (_data) => {
    const { title, body } = _data;
    const data = { title, body };
    this.props.createBlog(data)
      .then((res) => {
        console.log(res.success);
        history.push('/user');
      })
      .catch(err => console.log(err));
  }

  // If edit an existed blog
  handleEdit = (data) => {
    this.props.updateBlog(data)
      .then((res) => {
        console.log(res.success);
        history.push('/user');
      })
      .catch(err => console.log(err));
  }

  // Check if input is valid
  validateInput = () => {
    const { title, body } = this.state;
    if (title.length < constants.BLOG_TITLE_LENGTH_MIN
      || title.length > constants.BLOG_TITLE_LENGTH_MAX
      || body.length < constants.BLOG_BODY_LENGTH_MIN
      || body.length > constants.BLOG_BODY_LENGTH_MAX) {
      this.setState({ error: true });
      return false;
    }
    return true;
  }

  // Submit form
  onSubmit = () => {
    if (!this.validateInput()) {
      return false;
    }
    const { id, title, body, type } = this.state;
    const data = { id, title, body };
    if (type === 'edit') {
      this.handleEdit(data);
    } else if (type === 'new') {
      this.handleNew(data);
    }
    return true;
  }

  fetchBlog = (id) => {
    this.props.getEdittingBlog(id)
      .then((res) => {
        if (!res.success) return;
        const { edittingBlog } = this.props;
        this.setState({
          isLoading: false,
          title: edittingBlog.title,
          body: edittingBlog.body,
        });
      })
      .catch(err => console.log(err));
  }

  renderEditor = () => {
    const { isLoading, body } = this.state;
    if (isLoading) {
      return null;
    }
    return (
      <ReactMde
        className="md-editor"
        onChange={this.handleBodyChange}
        value={body}
        generateMarkdownPreview={markdown => Promise.resolve(this.converter.makeHtml(markdown))}
      />
    );
  }

  // render error if input is not valid
  renderError = () => {
    const { error } = this.state;
    if (error) {
      return (
        <div style={{ color: 'red' }}>
          {`Title must be between ${constants.BLOG_TITLE_LENGTH_MIN}-${constants.BLOG_TITLE_LENGTH_MAX} characters.
          Body must be between ${constants.BLOG_BODY_LENGTH_MIN}-${constants.BLOG_BODY_LENGTH_MAX} characters`}
        </div>
      );
    }
    return null;
  }

  render() {
    const { title } = this.state;
    return (
      <div className="user-container">
        <Header />
        <div className="user-inner">
          <div className="editor-container">
            <TextField
              onChange={this.handleTitleChange}
              value={title}
              className="editor-title-input"
              placeholder="10 - 100 characters"
              required
              id="outlined-required"
              label="Title"
              margin="normal"
              variant="outlined"
            />

            <Button
              variant="contained"
              color="secondary"
              onClick={this.onSubmit}
            >
              Submit
            </Button>

            {this.renderError()}

            {this.renderEditor()}

          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ user }) => ({
  edittingBlog: user.edittingBlog,
});

const mapDispatchToProps = {
  createBlog,
  updateBlog,
  getEdittingBlog,
};

export default connect(mapStateToProps, mapDispatchToProps)(BlogEditor);