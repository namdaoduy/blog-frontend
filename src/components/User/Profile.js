import React, { Component } from 'react';
import { Typography, Button, Grid, Divider, Avatar } from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import DeleteForever from '@material-ui/icons/DeleteForever';
import BorderColor from '@material-ui/icons/BorderColor';
import Moment from 'react-moment';
import '../../assets/styles/user.css';
import { MuiThemeProvider } from '@material-ui/core/styles';
import Header from '../Common/Header';
import theme from '../../constants/theme';
import API from '../../services/apis';
import history from '../../utils/history';

export default class Profile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_id: localStorage.getItem('justblog_user_id') || null,
      blogs: [],
      user_info: {},
    };
  }

  componentDidMount() {
    const uid = localStorage.getItem('justblog_user_id') || null;
    if (uid) {
      this.setState({ user_id: uid }, () => {
        this.fetchUserInfo();
        setTimeout(() => {
          this.fetchUserBlogs();
        }, 1000);
      });
    }
  }

  fetchUserInfo = () => {
    const { user_id } = this.state;
    API.getUserInfo(user_id)
      .then((res) => {
        this.setState({ user_info: res.data });
      })
      .catch(err => console.log(err));
  }

  fetchUserBlogs = () => {
    const { user_id } = this.state;
    API.getBlogsByUser(user_id)
      .then((res) => {
        this.setState({ blogs: res.data });
      })
      .catch(err => console.log(err));
  }

  handleDelete = (blog_id) => {
    const del = window.confirm('Are you sure to delete?');
    if (!del) return;
    const token = localStorage.getItem('justblog_access_token');
    API.deleteBlogById(blog_id, token)
      .then(() => {
        window.location.reload();
      })
      .catch(err => console.log(err));
  }

  handleClickEdit = (id) => {
    history.push(`/user/blog/${id}`);
  }

  handleClickNew = () => {
    history.push('/user/blog/new');
  }

  renderBlogs = () => {
    const { blogs } = this.state;
    return blogs.slice(0).reverse().map((blog, i) => (
      <Paper className="user-profile-blog">
        <Typography variant="h5" className="serif-2">
          {blog.title || 'Blog title'}
        </Typography>
        <Typography variant="body1" noWrap>
          {blog.body || 'Blog body'}
        </Typography>
        <Typography variant="subtitle2">
          {blog.author || 'Blog author'}
        </Typography>
        <Typography variant="caption">
          <Moment fromNow>{blog.created_at || 1545813100264}</Moment>
          <span className="dot-divider" />
          {blog.timeRead || `${10} min read`}
        </Typography>
        <Button
          color="secondary"
          variant="fab"
          onClick={() => this.handleDelete(blog.id)}
        >
          <DeleteForever />
        </Button>
        <Button
          color="secondary"
          variant="fab"
          onClick={() => this.handleClickEdit(blog.id)}
        >
          <BorderColor />
        </Button>
      </Paper>
    ));
  }

  render() {
    const { user_info } = this.state;
    const { name, email, picture } = user_info;
    return (
      <MuiThemeProvider theme={theme}>
        <div className="user-container">
          <Header />
          <Grid
            className="user-inner"
            container
            spacing={24}
          >
            <Grid item xs={3}>
              <Avatar alt="User Avatar" className="user-avatar" src={picture || null} />
              <Typography variant="h5" className="bold">
                {name || 'User Name'}
              </Typography>
              <Typography variant="subtitle1">
                {email || 'useremail@gmail.com'}
              </Typography>
              <Button
                variant="contained"
                color="secondary"
                onClick={this.handleClickNew}
              >
                New Blog
              </Button>
            </Grid>

            <Grid item xs={9}>
              <Typography variant="h5" className="semi-bold serif-2" gutterBottom>
                Your Blogs
              </Typography>
              <Divider />
              {this.renderBlogs()}
            </Grid>
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}