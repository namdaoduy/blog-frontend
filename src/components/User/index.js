import React, { Component } from 'react';
import '../../assets/styles/user.css';
import { Switch, Route } from 'react-router-dom';
import { Grid } from '@material-ui/core';
import { MuiThemeProvider } from '@material-ui/core/styles';
import Header from '../Common/Header';
import theme from '../../constants/theme';
import Profile from './Profile';
import BlogEditor from './BlogEditor';
import API from '../../services/apis';

export default class User extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_id: localStorage.getItem('justblog_user_id'),
      blogs: [],
    };
  }

  componentDidMount() {
    this.fetchUserBlogs();
  }

  fetchUserBlogs = () => {
    API.getBlogsByUser(this.state.user_id)
      .then((res) => {
        this.setState({ blogs: res.data });
      });
  }

  handleDelete = (id) => {
    const del = window.confirm('Are you sure to delete?');
    if (!del) return;
    const token = localStorage.getItem('justblog_access_token');
    API.deleteBlogById(id, token)
      .then((res) => {
        window.location.reload();
      })
      .catch(err => console.log(err));
  }

  handleClickEdit = (id) => {
    this.props.history.push(`/user/blog/${id}`);
  }

  handleClickNew = () => {
    this.props.history.push('/user/blog/new');
  }

  handleNew = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.postBlog(data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  handleEdit = (data) => {
    const token = localStorage.getItem('justblog_access_token');
    API.putBlog(data.id, data.title, data.body, token)
      .then((res) => {
        console.log(res);
        this.props.history.push('/user');
      });
  }

  render() {
    return (
      <MuiThemeProvider theme={theme}>
        <div className="user-container">
          <Header />
          <Grid
            className="user-inner"
            container
            spacing={24}
          >
            <Switch>
              <Route
                path="/user/blog/new"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="new"
                    key={`blogNew${Date.now()}`}
                    handleNew={this.handleNew.bind(this)}
                  />
                )}
              />
              <Route
                path="/user/blog/:blog_id"
                render={props => (
                  <BlogEditor
                    {...props}
                    type="edit"
                    key={`blogEdit${Date.now()}`}
                    handleEdit={this.handleEdit.bind(this)}
                  />
                )}
              />
              <Route
                path="/user"
                render={props => (
                  <Profile
                    {...props}
                    key={`profile${Date.now()}`}
                    handleClickNew={this.handleClickNew.bind(this)}
                    handleClickEdit={this.handleClickEdit.bind(this)}
                    handleDelete={this.handleDelete}
                    handleEdit={this.handleEdit}
                    blogs={this.state.blogs}
                  />
                )}
              />
            </Switch>
          </Grid>
        </div>
      </MuiThemeProvider>
    );
  }
}