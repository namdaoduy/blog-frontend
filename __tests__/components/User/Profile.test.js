import React from 'react';
import { shallow } from 'enzyme';
import { Profile } from 'components/User/Profile';
import Auth from 'utils/auth';

jest.mock('utils/auth');

describe('components/User/Profile', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<Profile {...props} />);
  };

  afterAll(() => {
    Auth.mockClear();
  });

  beforeEach(() => {
    props = {
      userInfo: {},
      userBlogs: [{
        id: 1,
        title: 'Mock title',
        body: 'Mock body',
        author: 'Mock Author',
        created_at: 123456789,
      }],
      getUserInfo: jest.fn(async () => {}),
      getUserBlogs: jest.fn(async () => {}),
      deleteBlog: jest.fn(async () => {}),
    };
  });

  it('Render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('Life cycles', () => {
    it('componentDidmount call fetch API function', () => {
      setup();
      const instance = wrapper.instance();
      const mockInfo = jest.spyOn(instance, 'fetchUserInfo');
      const mockBlog = jest.spyOn(instance, 'fetchUserBlogs');
      instance.componentDidMount();
      expect(mockInfo).toHaveBeenCalledTimes(1);
      expect(mockBlog).toHaveBeenCalledTimes(1);
    });
  });
});