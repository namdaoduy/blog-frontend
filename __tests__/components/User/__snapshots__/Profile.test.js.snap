// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/User/Profile Render correctly 1`] = `
<div
  className="user-container"
>
  <Connect(Header) />
  <WithStyles(Grid)
    className="user-inner"
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      item={true}
      xs={3}
    >
      <WithStyles(Avatar)
        alt="User Avatar"
        className="user-avatar"
        src={null}
      />
      <WithStyles(Typography)
        className="bold"
        variant="h5"
      >
        User Name
      </WithStyles(Typography)>
      <WithStyles(Typography)
        variant="subtitle1"
      >
        useremail@gmail.com
      </WithStyles(Typography)>
      <WithStyles(Button)
        color="secondary"
        onClick={[Function]}
        variant="contained"
      >
        New Blog
      </WithStyles(Button)>
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={9}
    >
      <WithStyles(Typography)
        className="semi-bold serif-2"
        gutterBottom={true}
        variant="h5"
      >
        Your Blogs
      </WithStyles(Typography)>
      <WithStyles(Divider) />
      <WithStyles(Paper)
        className="user-profile-blog"
        key="1"
      >
        <WithStyles(Typography)
          className="serif-2"
          variant="h5"
        >
          Mock title
        </WithStyles(Typography)>
        <WithStyles(Typography)
          noWrap={true}
          variant="body1"
        >
          Mock body
        </WithStyles(Typography)>
        <WithStyles(Typography)
          variant="subtitle2"
        >
          Mock Author
        </WithStyles(Typography)>
        <WithStyles(Typography)
          variant="caption"
        >
          <t
            ago={false}
            calendar={false}
            decimal={false}
            element="time"
            filter={[Function]}
            fromNow={true}
            interval={60000}
            onChange={[Function]}
            titleFormat=""
            toNow={false}
            unit={null}
            unix={false}
            utc={false}
            withTitle={false}
          >
            123456789
          </t>
          <span
            className="dot-divider"
          />
          10 min read
        </WithStyles(Typography)>
        <WithStyles(Button)
          color="secondary"
          onClick={[Function]}
          variant="fab"
        >
          <pure(DeleteForeverIcon) />
        </WithStyles(Button)>
        <WithStyles(Button)
          color="secondary"
          onClick={[Function]}
          variant="fab"
        >
          <pure(BorderColorIcon) />
        </WithStyles(Button)>
      </WithStyles(Paper)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</div>
`;
